### DISCUSSION TITLE
analytic orbit response matrix change according to radiation state of lattice

### DISCUSSION THREAD

**Discussion by simoneliuzzo** [atcollab/at#482] [Python AT]
with radiation_on AT tracking returns a response matrix that is almost identical to the analytic formula:
<img width="375" alt="image" src="https://user-images.githubusercontent.com/30342742/189087742-6012a4d1-9ba0-44e6-a320-6439ea1b6420.png">

with radiation off AT tracking returns a response matrix that ignores the dispersion term in the formula above

The code to obtain numeric=analytic response matrices becomes (NOTICE RAD_STATE): 
```

 _MH[i][j] = - ( (opt_bpm[i].beta[0] * opt_cor[j].beta[0]) ** 0.5 ) / 2 / math.sin(math.pi*tunes[0]) *\
                           math.cos( abs(opt_cor[j].mu[0] - opt_bpm[i].mu[0]) - math.pi*tunes[0]) + \
                           rad_state*(opt_bpm[i].dispersion[0] * opt_cor[j].dispersion[0])/L0/alpha_c

_MV[i][j] = ( (opt_bpm[i].beta[1] * opt_cor[j].beta[1]) ** 0.5 ) / 2 / math.sin(math.pi*tunes[1]) *\
                           math.cos( abs(opt_cor[j].mu[1] - opt_bpm[i].mu[1]) - math.pi*tunes[1]) + \
                           rad_state*(opt_bpm[i].dispersion[2] * opt_cor[j].dispersion[2])/L0/alpha_c
```

The rad_state flag may be should actually be a cavity_state flag. I could not test, since ring.set_cavity has no impact on beam tracking when radiation is off (cavity pass method is IdentityPass).

I find this very risky for use in operation of response matrices computed by pyAT. 

If the lattice used to compute such matrices even numerically has radiation off, the dispersive term will be ignored. 

In fact just having the cavity on would be sufficent to recover a correct response matrix. 
However, set_cavity() has no impact on the lattice if radiation is off. 

I think that having the possibility to set cavity on and radiation off is useful in this case.

**Comment by lfarv** [comment]
@simoneliuzzo:
You are perfectly right. The dispersive term occurs as soon as there is dispersion at the steerer location.
Setting cavities only is very easy, and even more with #453 (still waiting for @swhite2401's approval). Please have a look :
Matlab:
```Matlab
newring=atradon(ring,allpass='',cavitypass='auto');
```
python:
```python
newring=ring.radiation_on(all_pass=None, cavity_pass='auto', copy=True)
```

You will also note that there are now aliases for the "badly" named functions and properties:

`atradon` <=> `atenable6d`
`atradoff` <=> `atdisable6d`
`atGetRingProperties(ring,'radiation')` <=> `atGetRingProperties(ring,'is_6d')`

`ring.radiation_on` <=> `ring.enable_6d`
`ring.radiation_off` <=> `ring.disable_6d`
`ring.radiation` <=> `ring.is_6d`

And a new property tells if there is an **active** cavity (RFCavityPass) in the ring (independently of radiation):
`atGetRingProperties(ring,'has_cavity')`
`ring.has_cavity`

I think that this pull request answers your question.
If you have comments on #453, it's still time !

**Reply by simoneliuzzo** [reply]
Dear @lfarv,

I would like to keep radiation off and still have the cavity set correctly. Is there a way to achieve this?

In particular I have lattice files (FCC 100km long ring) where radiation is really to be kept off as it is a big complication (magnets gradients have to change with radiation on.

**Reply by lfarv** [reply]
@simoneliuzzo : exactly what I showed. Starting with radiation off:
```Matlab
% Turn cavities on
newring=atradon(ring, 'RFCavityPass', '', '', wigglerpass=''); % now
newring=atradon(ring, cavitypass='auto', allpass=''); % with #453 

% Set nominal frequency
newring=atsetcavity(newring, Frequency='nominal');
```
will turn cavities on, but nothing else => no radiation.

**Reply by simoneliuzzo** [reply]
Dear @lfarv,

Is this possible also in pyAT?

thank you
best regards
Simone

**Reply by simoneliuzzo** [reply]
`newring=ring.radiation_on(all_pass=None, cavity_pass='auto', copy=True)`
will turn cavity on and radiation off?

**Reply by lfarv** [reply]
Of course yes ! Be careful: not turn off, but leave unchanged ! **You have to start from a ring with everything off** !

`all_pass` does not exist until #453 is merged, so at the moment you have to type:
```
newring=ring.radiation_on(cavity_pass='RFCavityPass', dipole_pass=None, quadrupole_pass=None,
    wiggler_pass=None, copy=True)
```

`atradon` and `ring.radiation_on` always allowed to turn on selectively what you want !

**Reply by simoneliuzzo** [reply]
Dear @lfarv,

I must say that it is not really intuitive.
May be there is a plan for additional aliases to make these commands do what they say?
 
In this example, I would like to turn radiation OFF and cavity ON. The command are:

```
ring.radiation_off()
newring=ring.radiation_on(all_pass=None, cavity_pass='auto', copy=True)
```
To turn the cavity ON I have to use the command `radiation_on` (but I wanted it OFF) and ask for `cavity_pass='auto'` (not ON).

Also, I usually do not know what I start from, I would like to set the states of radiation and cavity independently from their original state.

It would be nice to have some properties in pyAT like `ring.radiation` to say if the cavity is active (RFCavityPass) or not, for example `ring.is_cavity_active` and the corresponding `ring.set_cavity_on()` and `ring.set_cavity_off()`. 

May be this could be part of #453 ?

**Reply by lfarv** [reply]
> I must say that it is not really intuitive.
> May be there is a plan for additional aliases to make these commands do what they say?

`radiation_on` turns things ON, for any specified things,
`radiation_off` turns things OFF, for any specified things.
How can it be more clear ?

> In this example, I would like to turn radiation OFF and cavity ON. The command are:
> ```
> ring.radiation_off()
> newring=ring.radiation_on(all_pass=None, cavity_pass='auto', copy=True)
> ```
> To turn the cavity ON I have to use the command `radiation_on` (but I wanted it OFF)

Agreed. As already mentioned, the alias `enable_6d` is now strongly encouraged !

> and ask for `cavity_pass='auto'` (not ON).

Because there are different ways to turn ON:
- 'SomethingPass': lets you explicitly specify the pass method
- 'auto': use the default pass method
- '' (or `None` in python): Don't touch! It's how you specify things you want to turn ON and ignore the rest.

> Also, I usually do not know what I start from, I would like to set the states of radiation and cavity independently from their original state.

Easy:
```ring.radiation_off(cavity_pass=None)```
Turn radiation OFF and keep the cavity unchanged.
```ring.radiation_on(all_pass=None, cavity_pass='auto')```
Turn cavity ON and leave the rest unchanged

> It would be nice to have some properties in pyAT like ring.radiation to say if the cavity is active (RFCavityPass) or not

As mentioned above, there is the `has_cavity` property which tells you that.

> and the corresponding ring.set_cavity_on()

`ring.radiation_on(all_pass=None, cavity_pass='auto')` does exactly that: Set the cavity and ignore all the rest.
Having a dedicated function for cavities would imply creating `set_dipole_on()`,  `set_quadrupole_on()`... functions, and continue creating new ones for each new element…

**Comment by lfarv** [comment]
@simoneliuzzo: Let me explain how we came to the present definition in #453:
I'm sorry that you were not implied from the beginning… It was initiated by @swhite2401 as a very simple bug fix (problem with QuantumDiffusion), but it turned later into a major upgrade ! I encourage you to go through the lengthy discussion of this PR, but here is a summary:

- What matters for linear optics is the presence of longitudinal motion or not: if none, the standard linear optics (&beta;-functions, dispersion…) stands, otherwise not. This motion may come from RF cavities, synchrotron radiation, longitudinal kicker, quantum diffusion element, impedance elements or possible new elements… We went for a single function to handle them all, rather that multiplying the functions. Obviously, the "historical" name `radiation_on` is misleading. It cannot be changed any more, but we provide an alias: `enable_6d`.
  It could have been `enable_longitudinal_motion`, or `enable_momentum_variation`…, but these are really long.
  Additional dedicated functions may be added later, as mentioned in https://github.com/atcollab/at/pull/453#issuecomment-1209062817.
- We went for 2 separate functions for enabling and disabling, rather than a single one like `set_6d`:
  - a single function would require a more complex syntax,
  - historically it's been like that from the beginning,
  - more important, as explained in the next point: with a single function, the default behaviour (empty argument list) is ambiguous.
- We want the default behaviour (empty argument list) to correspond to the most usual case. The present choices results from 2 modifications, and is:
  - `disable_6d()`: Turn **everything** OFF
  - `enable_6d()`: Turn ON cavities, set synchrotron radiation in dipoles, quadrupoles and wigglers.
  
  This is not symmetrical, but other choices open the risk of having unexpected elements turned on.

To summarise, here is the list of the new namings (they were already discussed): https://github.com/atcollab/at/pull/453#issuecomment-1209666408

`atradon` <=> `atenable6d`
`atradoff` <=> `atdisable6d`
`atGetRingProperties(ring,'radiation')` <=> `atGetRingProperties(ring,'is_6d')`
'check_radiation` <=> `check_6d`

`ring.radiation_on` <=> `ring.enable_6d`
`ring.radiation_off` <=> `ring.disable_6d`
`ring.radiation` <=> `ring.is_6d`
`@check_6d` decorator: no alias up to now

New property telling if there is an **active** cavity (RFCavityPass) in the ring (independently of radiation):
`atGetRingProperties(ring,'has_cavity')`
`ring.has_cavity`

The help of the legacy functions says that they are deprecated and should  be replaced by the more explicit new ones, but of course they stay unchanged.

**Reply by swhite2401** [reply]
@lfarv, `ring.has_cavity` does not seem to exist after pulling the latest master, also I could not find in PR #453 . Am I missing something?

**Reply by lfarv** [reply]
You are right, and you did not miss anything ! Three things are missing, as a result of some confusion in my local branches:

- `Lattice.has_cavity`
- the `check_6d` and `set_6d` decorators (alias for `check_radiation` and `set_radiation`)

I'll correct that in a next PR.

**Comment by lfarv** [comment]
@simoneliuzzo:

In python, `Lattice.enable_6d` and` Lattice.disable_6d` have a new alternative syntax that you may find simpler.
- this syntax applies only to the new functions, not to `radiation_on/off`
- this is unfortunately not available in Matlab (`atenable6d`)

Instead of using keywords like `cavity_pass=*`, you use positional elements. **You cannot mix both**.
The positional elements are Element class names, and the function will enable/disable all instances of these classes.
New classes are added to the base classes of some elements, allowing to select groups of elements:
- `Radiative` is a base class for Dipoles, Quadrupoles and Wigglers,
- `Collective` is a base class for all collective effects (even those are transverse only…).

#### Examples:
1. `ring.enable_6d(RFCavity)` turns all cavities on, so your example may be written:
   ```
   ring.disable_6d()
   newring = ring.enable_6d(RFCavity, copy=True)
   ```
2. `ring.enable_6d(RFCavity, Radiative)` is equivalent to the default behaviour `ring.enable_6d()`: turn on cavities, dipoles, quadrupoles and wigglers,
3. `ring.disable_6d(Collective)` will disable all impedance elements

With this new syntax, dedicated functions are less useful (at least in python).
